Sources/Foundation/JSONSerialization.swift

```diff
@@ -176,33 +179,94 @@ open class JSONSerialization : NSObject {
        The data must be in one of the 5 supported encodings listed in the JSON specification: UTF-8, UTF-16LE, UTF-16BE, UTF-32LE, UTF-32BE. The data may or may not have a BOM. The most efficient encoding to use for parsing is UTF-8, so if you have a choice in encoding the data passed to this method, use UTF-8.
      */
     open class func jsonObject(with data: Data, options opt: ReadingOptions = []) throws -> Any {
-        return try data.withUnsafeBytes { (rawBuffer: UnsafeRawBufferPointer) -> Any in
-            let encoding: String.Encoding
-            let bytes = rawBuffer.baseAddress!.assumingMemoryBound(to: UInt8.self)
-            let buffer: UnsafeBufferPointer<UInt8>
-            if let detected = parseBOM(bytes, length: data.count) {
-                encoding = detected.encoding
-                buffer = UnsafeBufferPointer(start: bytes.advanced(by: detected.skipLength), count: data.count - detected.skipLength)
-            }
-            else {
-                encoding = detectEncoding(bytes, data.count)
-                buffer = UnsafeBufferPointer(start: bytes, count: data.count)
+        do {
+            let jsonValue = try data.withUnsafeBytes { (ptr) -> JSONValue in
+                // first parse boom
+                var encoding: String.Encoding? = nil
+                if let (encoding, advanceBy) = parseBOM(ptr) {
+                    let newPtr = ptr[advanceBy..<ptr.count]
+                    let json = String(bytes: newPtr, encoding: encoding)!
+                    return try json.utf8.withContiguousStorageIfAvailable { (utf8) -> JSONValue in
+                        try JSONParser().parse(bytes: utf8)
+                    }!
+                }
+                
+                // second try to detect encoding in another way
+                if encoding == nil {
+                    encoding = JSONSerialization.detectEncoding(ptr)
+                }
+                
+                // if utf8 all is good
+                if encoding == .utf8 {
+                    return try JSONParser().parse(bytes: ptr)
+                }
+                
+                #warning("@Fabian: pretty sure we should throw an error here, if this is invalid data")
+                let json = String(bytes: ptr, encoding: encoding!)!
+                return try json.utf8.withContiguousStorageIfAvailable { (utf8) -> JSONValue in
+                    try JSONParser().parse(bytes: utf8)
+                }!
```

Comment:
Can this ever fail? Do we need explicit error handling here? @drexin @weissi

Reply:
On Darwin, this can fail. On Linux it cannot.

You can write

```
var json = String(bytes:...)
json.makeContiguousUTF8()
```

and after that you can `!` it.

Reply:
isn't json always utf8 though?

Reply:
I found a failure input which is UTF8 with BOM and bad data, Here is a test case to add to 
`TestJSONSerialization.test_JSONObjectWithData_encodingDetection`:

`("{} UTF-8 w/BOM w/ trailing illegal data", [0xEF, 0xBB, 0xBF, 0x7B, 0x7D, 0xff, 0x00]),`
(will need a test to confirm the JSON parsing failed and returned `nil`)


I would suggest merging (or at least have one call the other) `parseBOM` and `detectEncoding` - I cant see anywhere else they are used. If the encoding is determined to be UTF-8 then any leading BOM can be skipped as a new `String` does not need to be created. If if looks like UTF-8 a quick check if the first character is ASCII should be sufficient conformation as I dont think any JSON can start with a unicode point > 127.

Then the code could be simplified to something like:
```

guard let (encoding, advanceBy) = parseBOM(ptr) else { throw ... }
let newPtr = ptr[advanceBy..<ptr.count]
if encoding == .utf8 {
     parse it from newPtr...
 } else {
  guard let string = String(newPtr..) else { throw .. }
}
```

I couldnt find anything on json.org to say that JSON was guaranteed to be UTF-8 but its unlikely not to be so just converting to a new `String` as you have done is fine.

I would also avoid any force unwrapping when using `String()` and just `throw` since you are dealing with external

Reply:
from [here](https://tools.ietf.org/html/rfc4627)

```
3.  Encoding

   JSON text SHALL be encoded in Unicode.  The default encoding is
   UTF-8.

   Since the first two characters of a JSON text will always be ASCII
   characters [RFC0020], it is possible to determine whether an octet
   stream is UTF-8, UTF-16 (BE or LE), or UTF-32 (BE or LE) by looking
   at the pattern of nulls in the first four octets.

           00 00 00 xx  UTF-32BE
           00 xx 00 xx  UTF-16BE
           xx 00 00 00  UTF-32LE
           xx 00 xx 00  UTF-16LE
           xx xx xx xx  UTF-8
```

Reply:
For what it’s with, Darwin Foundation checks to make sure it’s UTF8 first and if it’s not just converts it to UTF8 for parsing.

Reply:
@parkera I'm afraid you quoted RFC 4672, which has been obsoleted for quite some time by 8258:

```
   JSON text exchanged between systems that are not part of a closed
   ecosystem MUST be encoded using UTF-8 [RFC3629].

   Previous specifications of JSON have not required the use of UTF-8
   when transmitting JSON text.  However, the vast majority of JSON-
   based software implementations have chosen to use the UTF-8 encoding,
   to the extent that it is the only encoding that achieves
   interoperability.

   Implementations MUST NOT add a byte order mark (U+FEFF) to the
   beginning of a networked-transmitted JSON text.  In the interests of
   interoperability, implementations that parse JSON texts MAY ignore
   the presence of a byte order mark rather than treating it as an
   error.
```

Am I right in assuming that the Darwin implementation allows UTF16 and UTF32 encoding and further allows byte order marks?

Reply:
I think the important takeaway from the 2 RFCs is that the input COULD be in any Unicode encoding but the output MUST always be UTF-8. However since the chance of the input being non-UTF8 is low the conversion from other inputs to UTF-8 as a one-time operation at the start is fine. Your code currently does all of this.

Running the test `test_JSONObjectWithData_encodingDetection` on Darwin using `DarwinCompatibilityTests` seems to indicate that Darwin doesn't handle some inputs (UTF16BE/LE witout BOM and UTF32LE w/BOM)